sql/core/src/main/scala/org/apache/spark/sql/execution/datasources/csv/CSVUtils.scala

```diff
@@ -97,23 +97,22 @@ object CSVUtils {
    */
   @throws[IllegalArgumentException]
   def toChar(str: String): Char = {
-    if (str.charAt(0) == '\\') {
-      str.charAt(1)
-      match {
-        case 't' => '\t'
-        case 'r' => '\r'
-        case 'b' => '\b'
-        case 'f' => '\f'
-        case '\"' => '\"' // In case user changes quote char and uses \" as delimiter in options
-        case '\'' => '\''
-        case 'u' if str == """\u0000""" => '\u0000'
-        case _ =>
-          throw new IllegalArgumentException(s"Unsupported special character for delimiter: $str")
-      }
-    } else if (str.length == 1) {
-      str.charAt(0)
-    } else {
-      throw new IllegalArgumentException(s"Delimiter cannot be more than one character: $str")
+    (str: Seq[Char]) match {
+      case Seq() => throw new IllegalArgumentException("Delimiter cannot be empty string")
+      case Seq(c) => c
```

Comment:
I'm missing why we had to switch up the case statement like this. I get that we need to cover more cases, but there was duplication and now there is a bit more. What about ...

```
str.length match {
  case 0 => // error
  case 1 => str(0)
  case 2 if str(0) == '\\' =>
    str(1) match {
      case c if """trbf"'\""".contains(c) => c
      case 'u' if str == """\u0000""" => '\0'
      case _ => // error
    }
  case _ => // error
}
```

Reply:
Ah yeah good points. This is too clever, off the top of my head. I still wonder if the code here can reduce the duplication of `Seq('\\', c) => '\c'` but I don't see a way that actually works, yeah.